### DISCUSSION TITLE
How to adjust Block2's memory limiting?

### DISCUSSION THREAD

**Discussion by jtcantin** [block-hczhai/block2-preview#84] [Q&A] [answered]
I have been trying to run some larger ground state energy calculations (transition metal organometallics) and I am getting the message `exceeding allowed memory`, along with the max size from Block2.

I looked at the debugging hints and there is something mentioned for the "Linear" condition, though it seems to be connected to molecule point groups (which I am not using in my calculations).

I have tried increasing both `stack_mem` and `stack_mem_ratio`, but the limit seems to be either unchanged or even smaller. That is, for default settings `stack_mem=1073741824` and `stack_mem_ratio=0.4`, I get `exceeding allowed memory (size=72477573, trying to allocate 157421)  (double)`. When I set `stack_mem=5368709120` (5*default) and `stack_mem_ratio=0.9`, I get `exceeding allowed memory (size=60397978, trying to allocate 127866)  (double)`. 

I have 202GB on the node and when I use python's `psutil` module, it indicates the node's memory is not being used above 15%. The queuing system indicates about 23GB max usage. I've also confirmed that `driver.stack_mem` and `driver.stack_mem_ratio` are set as mentioned above.

Are there other ways to increase the memory limits imposed by Block2 besides `stack_mem` and `stack_mem_ratio`?

**Comment by hczhai** [comment, marked as answer]
Thanks for reporting the problem.

If you look at the [API documentation](https://block2.readthedocs.io/en/latest/api/pyblock2.html#pyblock2.driver.core.DMRGDriver.__init__), for ``stack_mem`` it says: 

> The memory used for storing renormalized operators (in bytes). Default is 1 GB.

Note that ``1073741824 = 1024 * 1024 * 1024``.

In another [section](https://block2.readthedocs.io/en/latest/tutorial/qc-hamiltonians.html#Spin-Restricted-Integrals) in the documentation you can find that:

> For medium to large scale DMRG calculations, it is highly recommended to use a scratch space with high IO speed rather than the ``./tmp`` used in the following example. One also needs to set a suitable ``stack_mem`` in the ``DMRGDriver`` constructor to set the memory used for storing renormalized operators (in bytes). The default is ``stack_mem=int(1024**3)`` (1 GB). For medium scale calculations 10 to 30 GB might be required.

If you have 202 GB (= 202 * 1024 * 1024 * 1024 bytes) in the node, you may set ``stack_mem`` to ``214748364800 = 200 * 1024 * 1024 * 1024`` or something similar.

**Reply by jtcantin** [reply]
Thank you!
Putting `stack_mem` to a much larger value worked.

As I will likely want to optimize memory use to an extent, could you elaborate on what `stack_mem_ratio` does? It is defined as `The fraction of stack space occupied by the main stacks.`
What are the main stacks and what is stored in them?
Also, what is the memory not occupied by the main stacks used for?

**Reply by hczhai** [reply]
Both ``stack_mem`` and ``stack_mem_ratio`` are related to the max allowed memory usage. So they will not affect the actual memory usage. If you need to reduce the memory usage per node you may use the MPI version of ``block2`` and run the calculation in multiple nodes.
